### PR TITLE
feat(messaging): TenantId first-class envelope property (#228)

### LINKED ISSUE #228: Messaging Phase 1 / U2 — Promote TenantId to first-class envelope
Part of #217 — Messaging Phase 1 epic. See #217 for the full problem frame, requirements, decisions, and security model.

> **Scope note (Phase 1 reconciliation):** This unit ships in Phase 1 as part of the additive subset. The original `Dependencies: U1` reference is deferred — the publisher rename (U1) ships separately in Phase 2. U2 is independently valuable: tenant safety and envelope visibility do not require the rename.

---

**Goal:** Add typed `TenantId` to `PublishOptions` and `ConsumeContext`, add `Headers.TenantId` constant, and wire header↔property mapping in the shared publisher/consumer base.

**Requirements:** R4.

**Dependencies:** U1.

**Files:**
- Modify: `src/Headless.Messaging.Abstractions/PublishOptions.cs` (add `string? TenantId`)
- Modify: `src/Headless.Messaging.Abstractions/ConsumeContext.cs` (add `string? TenantId`, `DeliveryKind` property)
- Modify: `src/Headless.Messaging.Abstractions/Headers.cs` (add `TenantId = "headless-tenant-id"`)
- Modify: `src/Headless.Messaging.Core/` shared publish/consume pipeline (map property ↔ header)
- Test: `tests/Headless.Messaging.Tests.Unit/Abstractions/TenantIdRoundTripTests.cs`

**Approach:**
- On publish: the pipeline reconciles `PublishOptions.TenantId` and `headers["headless-tenant-id"]` with **fail-fast on disagreement**:
  - Property set, raw header unset → pipeline writes the property value to the raw header.
  - Property unset, raw header set → publish is **rejected** with `InvalidOperationException("TenantId raw header set without typed PublishOptions.TenantId — typed property is the authoritative seam; do not write the raw header directly.")`. Rationale: the raw header is reserved for transport-internal use; consumer code that bypasses the typed property is a strong indicator of a bug or an exfiltration attempt.
  - Both set, equal → no-op.
  - Both set, **disagree** → publish is **rejected** with `InvalidOperationException("TenantId property and raw header disagree. property={prop} header={hdr}")`. Silent overwrite was the prior behavior; it is replaced with a hard failure to surface header-injection attempts at publish time rather than letting a forged header round-trip through the outbox row, the OTel `headless.messaging.tenant_id` tag, and any downstream sink that consumes either.
- **Trust model (carries forward from R4 with a sharper boundary):** `TenantId` carries no authenticity guarantee end-to-end — a malicious or misconfigured publisher on a shared broker can still write any value into the typed property. The publish-time mismatch check above defends against the *narrow* case where attacker code can set the raw header dictionary but not the typed property (e.g., a low-privilege middleware injecting `headers`). Apps requiring full cross-tenant authenticity layer their own check via an `IConsumeBehavior<T>` in Phase 2 that validates `ConsumeContext.TenantId` against an out-of-band signal (mTLS SAN, signed header, channel binding). Consumer apps with a multi-tenant trust boundary MUST NOT trust `ConsumeContext.TenantId` alone for authorization decisions; this is documented as a callout in U6.
- On consume: the pipeline reads `headers["headless-tenant-id"]` and populates `ConsumeContext.TenantId`. Missing header → property is `null`. The consume-side header value is bounds-checked to the same 200-char limit and character-set rules applied on publish (via `Headless.Checks`); a header that exceeds the limit or carries disallowed characters is treated as a malformed envelope — the message is rejected (logged at warning, surfaced to `IDeadLetterObserver` per U8) rather than silently truncated. This blocks an upstream sender on a shared broker from forging an oversized `headless-tenant-id` value that would propagate into the outbox `TenantId` column or OTel spans.
- Validate `TenantId` length (match `MessageId` 200-char limit for symmetry) via `Headless.Checks` at the publisher boundary.

**Patterns to follow:**
- `MessageId` / `CorrelationId` handling in `PublishOptions` — tenancy mirrors that pattern exactly.
- `Headless.Checks` `Argument.*` validation per `CLAUDE.md`.

**Test scenarios:**
- Happy path: Setting `PublishOptions.TenantId = "acme"` produces `Headers["headless-tenant-id"] = "acme"` on the wire fake.
- Happy path: A consume pipeline wrapping a fake transport with `headers["headless-tenant-id"] = "acme"` yields `ConsumeContext.TenantId == "acme"`.
- Edge case: Missing tenant header → `ConsumeContext.TenantId` is `null`, not empty string.
- Edge case: Caller sets both property and raw header with different values → property wins, debug warning logged.
- Error path: `TenantId` exceeds 200 chars → `Argument.*` throws.

**Verification:**
- Round-trip test passes for both Send and Broadcast pipelines.
- No transport-specific code parses the tenant header directly — all consumers go through `ConsumeContext.TenantId`.

**Rollout-state callout (U2 ⇄ U3a/U3b):** U2 introduces the envelope contract (property + header + shared pipeline). Per-provider dedup-key and outbox unique-constraint migrations to composite `(TenantId, MessageId)` land inside U3a and U3b. Interim state between U2 and full U3a/U3b completion: `TenantId` is envelope-visible on all providers but dedup-correct only on providers whose migration has landed. The U6 capability matrix tracks per-provider dedup-migration status until every provider is green.

**Issue acceptance criteria**
- [ ] All test scenarios listed above pass.
- [ ] XML docs updated on every modified public type (`PublishOptions`, `ConsumeContext`, `Headers`).
- [ ] No transport-specific code parses the tenant header directly — confirmed via `grep -r '"headless-tenant-id"' src/Headless.Messaging.* | grep -v Headers.cs` returning zero results.

### PR DESCRIPTION
**Pull request by xshaheen**
## Summary

Promotes `TenantId` from an informal header convention to a typed first-class envelope property on `PublishOptions` and `ConsumeContext`, with a strict 4-case publish-time integrity policy centralized in `Headless.Messaging.Core` so all 11 transports inherit it without per-provider duplication.

Closes #228. Phase 1/U2 of the Headless.Messaging epic (#217).

## What's new

**Abstractions surface (one new constant, two new init-set properties):**

- `Headers.TenantId = "headless-tenant-id"` — canonical wire-header key.
- `PublishOptions.TenantId : string?` + `PublishOptions.TenantIdMaxLength = 200`.
- `ConsumeContext<T>.TenantId : string?` (init, **not** `required`, mirrors `CorrelationId`).

**Publish-time integrity (`MessagePublishRequestFactory._ApplyTenantId`):**

| typed | raw | outcome |
|---|---|---|
| set | unset | stamp header from typed value |
| unset | set | reject with `InvalidOperationException` (header reserved for typed surface) |
| both, equal | both, equal | emit (no-op reconciliation) |
| both, disagree | both, disagree | reject with `InvalidOperationException` carrying both values |

Whitespace raw headers are treated as unset to mirror the consume-side leniency. Whitespace-only typed values throw because `TenantId` has no auto-generated default. `Headers.TenantId` is intentionally **not** added to `_ReservedHeaders` (case "both equal" must accept an equal raw header).

**Consume-time leniency (`ConsumeExecutionPipeline._ResolveTenantId`):**

Missing, whitespace, or oversized (`> TenantIdMaxLength`) header values map to `null` rather than failing the message. Bound into the cached `_CompileFactory` MemberInit via `Expression.Call`, parallel to the existing `_ResolveTimestamp` helper.

## Why this shape

- **One seam, eleven transports** — every transport composes through `IMessagePublishRequestFactory.Create`. Centralizing the integrity policy here means all transports inherit it for free with no per-provider code.
- **Honors R7 (no new DI seams)** — extends the existing `_CreateHeaders` reserved-header pattern instead of introducing a decorator chain.
- **Symmetric publish/consume whitespace handling** — both sides treat empty/whitespace as "absent" to keep the rules predictable.
- **Documented but un-implementable test** — the R6 oversized-on-consume case is exercised by `_ResolveTenantId`'s direct logic. Adding a dedicated integration test would require `InternalsVisibleTo` on `Headless.Messaging.InMemoryQueue` (no public surface to inject a raw `TransportMessage`) or reflection on a `private static` helper, both of which violate project test conventions.

## Out of scope (from origin doc)

- `DeliveryKind` enum on `ConsumeContext` (deferred to U1/Phase 2)
- `AllowHeaderTenantHydration` opt-in (killed; strict policy is the resolved decision)
- `TenantContextRequired` startup validator (sibling follow-up issue, distinct failure semantics)
- Per-provider composite-key `(TenantId, MessageId)` dedup migration (Phase 2)
- Inbound envelope signing / cross-tenant authenticity (Phase 2)
- New exception types, options, or DI seams (R7)

## Test plan

- [x] `Headless.Messaging.Abstractions.Tests.Unit` — 62 tests pass (5 new `ConsumeContext.TenantId` accessor tests, 5 new `PublishOptionsTests` for the typed property)
- [x] `Headless.Messaging.Core.Tests.Unit / DirectPublisherTests` — 27 tests pass (9 new tests covering all 4 integrity cases plus null/oversized/whitespace edges)
- [x] `Headless.Messaging.Core.Tests.Unit / IDirectPublisherIntegrationTests` — 9 tests pass (2 new round-trip tests: set + unset, exercising publish→consume end-to-end)
- [x] No transport-specific code parses the wire header directly: `grep -r '"headless-tenant-id"' src/Headless.Messaging.* | grep -v Headers.cs` returns only doc comments and generated XML doc artifacts

## Coordination follow-ups (not in this PR)

- [x] Update issue #235 body to depend on #228 and use the typed property (`TenantPropagationOutboxPublisher` / `TenantPropagationConsumeFilter`) — done 2026-05-03
- [x] Open sibling strict-tenancy publish-guard issue (`MissingTenantContextException` family — distinct from #228's `InvalidOperationException`) — tracked in #238
- [x] Amend issue #217 to remove `AllowHeaderTenantHydration` references; note strict 4-case rejection as the resolved decision — #217 body now records the strict policy and child #228 is ticked

## Plan and brainstorm

- Brainstorm: `docs/brainstorms/2026-05-01-tenant-id-envelope-requirements.md`
- Plan: `docs/plans/2026-05-01-001-feat-tenant-id-envelope-plan.md`

### CODE REVIEW

File: src/Headless.Messaging.Core/Internal/IMessagePublishRequestFactory.cs [hunk 1]
```diff
@@ -116,6 +117,59 @@ private static string _ValidateMessageId(string messageId)
...
+        }
+
+        _ValidateTenantId(typed);
+
+        if (rawSet && !string.Equals(raw, typed, StringComparison.Ordinal))
+        {
+            throw new InvalidOperationException(
+                $"PublishOptions.TenantId='{typed}' disagrees with header '{Headers.TenantId}'='{raw}'. "
+                    + "Set the typed property only."
+            );
```
[the remark this answers is not in the document]
Comment (xshaheen) [reply src/Headless.Messaging.Core/Internal/IMessagePublishRequestFactory.cs:164]: > this can allow control characters/newlines into exception messages and logs. Consider escaping/sanitizing the values (or logging lengths/hashes) before including them in the exception message.

Addressed in `a1c72e89`. Both `_ApplyTenantId` exception paths now route the wire-side raw value through `LogSanitizer.Sanitize(raw, PublishOptions.TenantIdMaxLength)` (the same helper used by `ISubscribeExecutor`) before interpolation, so CR/LF/control chars cannot reach `Exception.Message` or downstream log sinks. Each exception also surfaces structured discriminators via `Exception.Data`: `Headless.Messaging.FailureCode` (`ReservedTenantHeader` for case b, `TenantIdMismatch` for case d), `Headers.TenantId.Raw` with the sanitized raw, and `PublishOptions.TenantId` with the typed value on the mismatch path.

File: src/Headless.Messaging.Core/Internal/IConsumeExecutionPipeline.cs [hunk 2]
```diff
@@ -224,6 +231,24 @@ private static Delegate _CompileFactory(Type messageType)
...
+    {
+        if (!headers.TryGetValue(Headers.TenantId, out var value) || string.IsNullOrWhiteSpace(value))
+        {
+            return null;
+        }
+
+        if (value.Length > PublishOptions.TenantIdMaxLength)
+        {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (xshaheen) [reply src/Headless.Messaging.Core/Internal/IConsumeExecutionPipeline.cs:247]: > The new consume-side rule "oversized (> TenantIdMaxLength) header maps to null" isn't covered by tests... consider exposing the resolver as internal (or otherwise adding a targeted unit/integration test) to lock in the oversized/whitespace mapping behavior.

Addressed in `40bfe7a5`. Three tests added to `tests/Headless.Messaging.Core.Tests.Unit/SubscribeInvokerTests.cs` exercising `_ResolveTenantId` end-to-end through the public `ISubscribeInvoker` surface (which drives `ConsumeExecutionPipeline._CompileFactory`):

- `should_populate_tenant_id_from_inbound_header` — happy path; `Headers.TenantId = "tenant-123"` flows verbatim to `ConsumeContext.TenantId`.
- `should_handle_max_length_tenant_id_on_consume` — boundary; 200-char value preserved.
- `should_handle_oversized_tenant_id_leniently_on_consume` — R6 lenient; 201-char value maps to `null`.

The existing `_CreateMediumMessage` helper gained an optional `tenantId` parameter, so the resolver is reachable from a public-API test path without changing visibility, adding `InternalsVisibleTo`, or using reflection. Class total: 9 → 12 tests, all green.

File: src/Headless.Messaging.Abstractions/Headers.cs [hunk 3]
```diff
@@ -58,6 +58,14 @@ public static class Headers
...
     public const string CallbackName = "headless-callback-name";
 
+    /// <summary>
+    /// Multi-tenancy identifier for the message, populated from <see cref="PublishOptions"/>.TenantId at publish time
+    /// and exposed on <see cref="ConsumeContext{TMessage}"/>.TenantId at consume time.
+    /// The publish pipeline rejects writes to this key via raw headers (use the typed property instead).
+    /// Value: "headless-tenant-id"
+    /// </summary>
+    public const string TenantId = "headless-tenant-id";
+
```
[the remark this answers is not in the document]
Comment (xshaheen) [reply src/Headless.Messaging.Abstractions/Headers.cs:80]: > the policy allows a raw header if it matches PublishOptions.TenantId (the "both equal" case). Consider clarifying that raw-only and mismatched raw headers are rejected, while matching raw+typed is accepted.

Addressed in `a1c72e89`. Rewrote the `Headers.TenantId` XML doc to spell out the full 4-case integrity policy in a `<remarks>` block: raw-only writes are rejected with `InvalidOperationException`, raw writes that disagree with the typed `PublishOptions.TenantId` are rejected, and a raw write that matches the typed property is accepted as a no-op reconciliation. Also fixed the cref syntax (`<see cref="PublishOptions.TenantId"/>` instead of `<see cref="PublishOptions"/>.TenantId`) and added an untrusted-wire-data warning for consume-side use.
